### PR TITLE
Remove `BitArray<T,SIZE>` in favor of `[Option<T>;SIZE]`

### PR DESCRIPTION
**Pull request by DelSkayn**
BitArray seemed to implement an array of optional values and used an extra array of booleans internally to keep track of which values where present and which were not. 

This could also be done more efficiently with just an array of options. Rust implements a optimization where options with non-nullable pointers inside of them have the same size as the plain pointer. This means that using an array of options is more space efficient.

There seemed to be one optimization in Node48 where the search for a empty spot might be faster over the bit array then over the array of pointers as the bit array uses less memory. To match that performance I used a bit field to mark which fields are used and which are empty. Using `trailing_ones` it is then possible to find an empty position in the array with a single instruction.

Overall this PR seemed to improve performance on the benchmarks when I ran them locally. Running `cargo criterion` resulted in the following results:

```
seq_insert/seq_insert   time:   [2.1321 µs 2.1716 µs 2.2037 µs]
                        thrpt:  [453.78 Kelem/s 460.50 Kelem/s 469.02 Kelem/s]
                 change:
                        time:   [-9.8536% +18.415% +71.873%] (p = 0.66 > 0.05)
                        thrpt:  [-41.817% -15.552% +10.931%]
                        No change in performance detected.

rand_insert/art         time:   [1.9651 µs 2.0091 µs 2.0440 µs]
                        thrpt:  [489.25 Kelem/s 497.73 Kelem/s 508.87 Kelem/s]
                 change:
                        time:   [-15.773% -11.318% -5.3067%] (p = 0.00 < 0.05)
                        thrpt:  [+5.6041% +12.762% +18.727%]
                        Performance has improved.

seq_get/art/1000000     time:   [17.452 ns 17.871 ns 18.413 ns]
                        thrpt:  [54.310 Melem/s 55.956 Melem/s 57.302 Melem/s]
                 change:
                        time:   [-61.570% -55.887% -49.153%] (p = 0.00 < 0.05)
                        thrpt:  [+96.668% +126.69% +160.21%]
                        Performance has improved.

random_get/art/1000000  time:   [214.45 ns 215.43 ns 216.26 ns]
                        thrpt:  [4.6241 Melem/s 4.6419 Melem/s 4.6632 Melem/s]
                 change:
                        time:   [-35.897% -34.513% -33.108%] (p = 0.00 < 0.05)
                        thrpt:  [+49.495% +52.702% +55.999%]
                        Performance has improved.

random_get_str/art/1000000
                        time:   [486.90 ns 490.07 ns 493.33 ns]
                        change: [-20.065% -17.987% -15.757%] (p = 0.00 < 0.05)
                        Performance has improved.
```

### CODE REVIEW

File: src/art.rs [hunk 1]
```diff
@@ -160,11 +160,11 @@ impl<P: KeyTrait + Clone, V: Clone> Node<P, V> {
     #[inline]
     fn is_full(&self) -> bool {
         match &self.node_type {
-            NodeType::Node1(n) => self.num_children() >= n.size(),
```
Comment (DelSkayn) [inline src/art.rs:163]: Going through `self.num_children()` does a match on the `node_type` again. So I changes this to just use the node directly.

File: src/node.rs [hunk 2]
```diff
@@ -369,35 +354,40 @@ impl<P: KeyTrait + Clone, N: Version, const WIDTH: usize> Drop for FlatNode<P, N
 pub struct Node48<P: KeyTrait + Clone, N: Version> {
     pub(crate) prefix: P,
     pub(crate) version: u64,
-    keys: BitArray<u8, 256>,
-    children: BitArray<Arc<N>, 48>,
-    num_children: u8,
+    keys: Box<[u8; 256]>,
+    children: Box<[Option<Arc<N>>; 48]>,
+    child_bitmap: u64,
```
Comment (DelSkayn) [inline src/node.rs:359]: This is a bit map keeping track of the zeros. Ones indicate that child is filled and zeroes indicate empty spots.

File: src/node.rs [hunk 3]
```diff
@@ -483,24 +483,27 @@ impl<P: KeyTrait + Clone, N: Version> NodeTrait<N> for Node48<P, N> {
...
+        let idx = self.keys[key as usize];
+        if idx == u8::MAX {
+            return None;
+        }
+        Some(self.children[idx as usize].as_ref().unwrap())
     }
 
     fn num_children(&self) -> usize {
-        self.num_children as usize
+        self.child_bitmap.count_ones() as usize
```
Comment (DelSkayn) [inline src/node.rs:506]: Since this field is now a bitmap the amount of children is the amount of ones in the integer.

File: src/art.rs (hunk 1 above)
Comment (arriqaaq) [reply src/art.rs:163]: awesome! Thanks for catching this

Comment (arriqaaq) [reply src/art.rs:163]: awesome!

**Review by arriqaaq — APPROVED** [review]

**Pull request merged**